src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs

```diff
@@ -0,0 +1,113 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matching;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    internal class DynamicControllerEndpointMatcherPolicy : MatcherPolicy, IEndpointSelectorPolicy
+    {
+        private readonly DynamicControllerEndpointSelector _selector;
+
+        public DynamicControllerEndpointMatcherPolicy(DynamicControllerEndpointSelector selector)
+        {
+            if (selector == null)
+            {
+                throw new ArgumentNullException(nameof(selector));
+            }
+
+            _selector = selector;
+        }
+
+        public override int Order => int.MinValue + 100;
+
+        public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            if (!ContainsDynamicEndpoints(endpoints))
+            {
+                // Dynamic controller endpoints are always dynamic endpoints.
+                return false;
+            }
+
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var metadata = endpoints[i].Metadata.GetMetadata<DynamicControllerMetadata>();
+                if (metadata != null)
+                {
+                    // Found a dynamic controller endpoint
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public async Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            // There's no real benefit here from trying to avoid the async state machine.
+            // We only execute on nodes that contain a dynamic policy, and thus always have
+            // to await something.
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                var endpoint = candidates[i].Endpoint;
+
+                var metadata = endpoint.Metadata.GetMetadata<DynamicControllerMetadata>();
+                if (metadata == null)
+                {
+                    continue;
+                }
+
+                var matchedValues = candidates[i].Values;
+                var endpoints = await _selector.SelectEndpointsAsync(metadata.Values);
```

Comment:
Does select throw when multiple endpoints get picked up?

Reply:
No, the documentation says that it's implementation defined. The reason why I made this implementation defined ordering instead of throwing is that we will have multiple matches in a trivial case - pages that have the filename `Index.cshtml`. This produces two endpoints. 

-----

We could change this depending on scenarios and feedback. Here are the concerns involved, I know @JamesNK and @davidfowl will be interested as well.

1. If you want this to throw - it breaks a common use case
2. If you want this to be implementation defined it's really simple to implement and explain, but the capabilities are pretty limited
3. If you want this to **allow** returning multiple endpoints, and have the rest of the system try to **disambiguate** them for you, then it's a very powerful and confusing system.

----

So let's assume that you want to be able to return multiple endpoints (3) and explore it a bit. Let's consider a basic case, like "I want to return two endpoints, and have HTTP method disambiguation select the best match".

Well, The *dynamic endpoint pattern* allows you to **RUN CODE** to replace the dynamic endpoint with the single endpoint of your choosing - **AFTER** the route pattern has been matched.

So I really need to add multiple endpoints but I can't because ....

This is a intentional limitation of `CandidateSet`. It's a data structure that's highly optimized for a small number of entries, and it doesn't allow you to insert items. We could change this, and it would be relatively inefficient everywhere you use it (relative to something like techempower).

Cool, so now I can insert multiple endpoints. The next problem is that ...

The matcher policy that you write to replace an endpoint runs in a pipeline of matcher policies. ie: some policies might run before, some might run after.  Well, if you want the system to disambiguate for you then you really are concerned about which policies run before you and which after. Policies that ran before you will see the dynamic endpoint, policies that run after you will see the replacement.

The next issue is that we have policies that bake themselves into the DFA. We've generally written our policies to use the DFA instead of running in the pipeline of matcher policies (HTTP method, Consumes, Hostname). This is like the difference between compiled (DFA) and interpreted (matcher policy). The DFA policy implementations become part of the core data structure of routing, but they cannot do this for a dynamic endpoint, because the metadata of that endpoint are unknown. We will need to make it possible for all of our policies to run in DFA model or `IEndpointSelectorPolicy` mode.

Now something like this could be possible.

If we're going to end up here - ie we need these features, then I think I would step back and make this a core feature of DFA matcher rather than a thing built on top.

Reply:
I like option 2 (until we know more or discuss more) because while it might not be convenient, it's not blocking either. Someone who wants to extend routing is taking on some *hard problems* - I don't really assume we can solve them for everyone.